### PR TITLE
feat: [290085] File download

### PR DESCRIPTION
**Pull request by KiranVarma02**

### CODE REVIEW

File: designer/server/src/routes/file/file.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      'string.empty': 'Enter an email address'
+    })
+})
+
+export default [
+  /**
+   * @satisfies {ServerRoute< { Params:  FileDownload, Payload: { email: string} }>}
+   */
+  ({
+    method: 'GET',
```
Comment (KiranVarma02) [inline designer/server/src/routes/file/file.js:26]: remove after testing

File: designer/server/src/routes/file/file.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    path: '/file-download/{fileId}',
+    async handler(request, h) {
```
Comment (KiranVarma02) [inline designer/server/src/routes/file/file.js:28]: remove after testing

File: designer/server/src/routes/file/file.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return Boom.internal(
+          new Error('Failed to download file', {
+            cause: err
+          })
+        )
+      }
+    },
+    options: {
+      auth: false
+    }
```
Comment (KiranVarma02) [inline designer/server/src/routes/file/file.js:55]: remove after testing

File: designer/server/src/routes/file/file.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    async handler(request, h) {
+      const { params, yar } = request
+      const { fileId } = params
+
+      const statusCode = await checkFileStatus(fileId)
+
+      switch (statusCode) {
+        case StatusCodes.OK: {
+          const validation = yar.flash(sessionNames.validationFailure)[0]
+          return h.view('file/download-page', file.fileViewModel(validation))
```
Comment (KiranVarma02) [inline designer/server/src/routes/file/file.js:32]: there is below tslint error on this line. Need help please 

`file.js(32,66): error TS2345: Argument of type 'ValidationFailure<FormMetadataInput>' is not assignable to parameter of type 'ValidationFailure<{ email: string; }>'.`

Comment (davidjamesstone) [reply designer/server/src/routes/file/file.js:32]: Have a look at the types for `yar` in `designer/server/src/types.ts` line 93. I think that's where the problem is. It doesn't know that `const validation = yar.flash(sessionNames.validationFailure)[0]` can return what you want it to in the next line `file.fileViewModel(validation)`.

Comment (KiranVarma02) [reply designer/server/src/routes/file/file.js:32]: I tried few things in https://github.com/DEFRA/forms-designer/pull/409/commits/20f00ddf43b2777b6290e5f7f32718591f9f708e but couldn't get it to work. I just defined as object which is not great, i will leave it to that unless someone has better suggestion.

File: designer/server/src/routes/file/file.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
+import Boom from '@hapi/boom'
+import { StatusCodes } from 'http-status-codes'
+import Joi from 'joi'
+
+import { redirectWithErrors } from '../forms/create.js'
```
[the remark this answers is not in the document]
Comment (KiranVarma02) [reply designer/server/src/routes/file/file.js:5]: done

File: designer/server/src/routes/file/file.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import { errorViewModel } from '~/src/models/errors.js'
+import * as file from '~/src/models/file/file.js'
+
+export const emailSchema = Joi.string().trim().required().messages({
+  'string.empty': 'Enter an email address'
+})
+
+export default [
+  /**
+   * @satisfies {ServerRoute< { Params:  { fileId: string } }>}
```
[the remark this answers is not in the document]
Comment (KiranVarma02) [reply designer/server/src/routes/file/file.js:18]: done

File: designer/server/src/routes/file/file.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        default: {
+          return h
+            .response('Unhandled file status')
+            .code(StatusCodes.INTERNAL_SERVER_ERROR)
+        }
+      }
+    }
+  }),
+  /**
+   * @satisfies {ServerRoute< { Params:  { fileId: string }, Payload: { email: string } }>}
```
[the remark this answers is not in the document]
Comment (KiranVarma02) [reply designer/server/src/routes/file/file.js:49]: done

File: designer/server/src/views/file/download-page.njk [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+{% from "govuk/components/input/macro.njk" import govukInput %}
+
+{% block content %}
+  <form method="post" novalidate>
+    {% call appPageBody({
+      heading: pageHeading,
+      errorList: errorList
+    }) %}
+
+    <p class="govuk-body-l">
```
Comment (davidjamesstone) [inline designer/server/src/views/file/download-page.njk:14]: This is `govuk-body` in the prototype. `govuk-body-l` looks a bit off.

**Comment by davidjamesstone** [comment]
image

Is this page out of scope? I can't see it.

https://documents.service.gov.uk/d/r5DUy66ISnyhl1wwx9tCOw/1bAr_WN3QNeQDZclJQr1Fw?key=UaOBRV0CwBHvXmumnqCAUGY26WmO_2sSenoN1imtkgY

**Review by davidjamesstone — APPROVED** [review]
I've loaded up you branch locally and all seems to work ok, Kiran. Just one final comment here and the question on the "conversation" tab.

Approved in advance. Thanks

**Comment by davidjamesstone** [comment]
Sorry Kiran. I've just noticed points 5, 6 and 9 on thi ticket.

```
5. Email address (password) becomes autofill and the user can confirm the selection or amend for additional downloads
6. Audit record to be created using the existing auditing approach / design to record the user who downloaded the file
This includes the FILE ID and not the file name
...
9. Audit record to be created using the existing auditing approach / design to record failed attempts to download file
```

Are these done?

File: designer/server/src/views/file/download-page.njk [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        isPageHeading: false
+      },
+      id: field.id,
+      name: field.name,
+      value: field.value,
+      errorMessage: {
+        text: formErrors[field.name].text
+      } if formErrors[field.name]
+      })
+    }}
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-page.njk:31]:
```suggestion
    }) }}
```

File: designer/server/src/views/file/download-page.njk (hunk 8 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-page.njk:14]: done

File: designer/server/src/views/file/download-page.njk (hunk 9 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-page.njk:31]: done

**Comment by KiranVarma02** [comment]
> 5. Email address (password) becomes autofill and the user can confirm the selection or amend for additional downloads

done except 5 as discussed

File: designer/server/src/views/file/download-complete.njk [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+{% block content %}
+  {% call appPageBody({
+    heading: pageHeading
+  }) %}
+    <p class="govuk-body-l">
+      Your file is downloading
+    </p>
+
+    <p class="govuk-body-l">
+      Once download is complete, your file will be saved in the 'Downloads' folder on your computer. If nothing happens, select 'Download file'.
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-complete.njk:19]:
```suggestion
      Once download is complete, your file will be saved in the ‘Downloads’ folder on your computer. If nothing happens, select ‘Download file’.
```

Comment (colinrotherham) [reply designer/server/src/views/file/download-complete.njk:19]: Should probably use the actual button text too

```suggestion
      Once download is complete, your file will be saved in the ‘Downloads’ folder on your computer. If nothing happens, select ‘{{ buttonText }}’.
```

File: designer/server/src/views/file/download-page.njk [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      id: field.id,
+      name: field.name,
+      value: field.value,
+      errorMessage: {
+        text: formErrors[field.name].text
+      } if formErrors[field.name]
+    }) }}
+
+    {{ govukButton({
+      text: 'Download file'
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-page.njk:33]: Should this be `buttonText` too?

```suggestion
      text: buttonText
```

File: designer/server/src/views/file/download-page.njk [hunk 12]
```diff
@@ -0,0 +1,37 @@
+{% extends 'layouts/page.njk' %}
+
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+{% from "govuk/components/input/macro.njk" import govukInput %}
+
+{% block content %}
+  <form method="post" novalidate>
+    {% call appPageBody({
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-page.njk:9]: Can you indent everything inside this call block by 2 spaces to match **download-complete.njk**?

File: designer/server/src/views/file/expired.njk [hunk 13]
```diff
@@ -0,0 +1,15 @@
+{% extends 'layouts/page.njk' %}
+
+{% block content %}
+  {% call appPageBody({
+    heading: pageHeading
+  }) %}
+    <p class="govuk-body-l">
```
Comment (davidjamesstone) [inline designer/server/src/views/file/expired.njk:7]: `govuk-body` not `govuk-body-l`

File: designer/server/src/views/file/expired.njk [hunk 14]
```diff
@@ -0,0 +1,15 @@
+{% extends 'layouts/page.njk' %}
```
Comment (colinrotherham) [inline designer/server/src/views/file/expired.njk:1]: A few have slipped in already, but can we stick to double quotes by default in Nunjucks?

```suggestion
{% extends "layouts/page.njk" %}
```

File: designer/server/src/views/file/download-page.njk [hunk 15]
```diff
@@ -0,0 +1,37 @@
+{% extends 'layouts/page.njk' %}
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-page.njk:1]:
```suggestion
{% extends "layouts/page.njk" %}
```

File: designer/server/src/views/file/download-complete.njk [hunk 16]
```diff
@@ -0,0 +1,28 @@
+{% extends 'layouts/page.njk' %}
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-complete.njk:1]:
```suggestion
{% extends "layouts/page.njk" %}
```

File: designer/server/src/routes/forms/create.js [hunk 17]
```diff
@@ -310,20 +310,22 @@ export function redirectToStepWithErrors(request, h, error) {
  * @param {ResponseToolkit} h
  * @param {Error} [error]
  * @param {boolean} [redirectToPreviousStep] Optionally redirect to errors on previous steps, else it uses the current URL
+ * @param {string} [flashKey] Optionally redirect to errors on previous steps, else it uses the current URL
```
Comment (colinrotherham) [inline designer/server/src/routes/forms/create.js:313]: Nice one @KiranVarma02

Just need to tell the compile that only known session keys are allowed

```suggestion
 * @param {keyof typeof sessionNames} [flashKey] Optionally redirect to errors on previous steps, else it uses the current URL
```

File: designer/server/src/views/file/download-complete.njk (hunk 10 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-complete.njk:16]: this page will most likely change but I have done above

File: designer/server/src/views/file/download-page.njk (hunk 15 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-page.njk:1]: done

File: designer/server/src/views/file/download-page.njk (hunk 12 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-page.njk:9]: done

File: designer/server/src/views/file/expired.njk (hunk 14 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/expired.njk:1]: done

File: designer/server/src/views/file/expired.njk (hunk 13 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/expired.njk:7]: done

File: designer/server/src/routes/forms/create.js (hunk 17 above)
Comment (KiranVarma02) [reply designer/server/src/routes/forms/create.js:313]: done

File: designer/server/src/views/file/download-complete.njk (hunk 16 above)
Comment (KiranVarma02) [reply designer/server/src/views/file/download-complete.njk]: done

**Review by alexluckett — APPROVED** [review]

File: designer/server/src/views/file/download-page.njk [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+{% block content %}
+  <form method="post" novalidate>
+    {% call appPageBody({
+      heading: pageHeading,
+      errorList: errorList
+    }) %}
+
+      <p class="govuk-body">
+          To confirm the file was meant for your team, enter the email address the file was sent to.
```
Comment (colinrotherham) [inline designer/server/src/views/file/download-page.njk:15]:
```suggestion
        To confirm the file was meant for your team, enter the email address the file was sent to.
```

**Review by colinrotherham — APPROVED** [review]
Looks good, I'm sure anything else will come out in testing

Did you mean to bring a merge commit in? This one: [`Merge remote-tracking branch 'origin/main' into feat/290085-file-down…`](https://github.com/DEFRA/forms-designer/pull/409/commits/f21e17a115fc3a54cc93fc07d41a74d3105ef836)

**Pull request merged**